PERCEPTION: Add buoy detection pcl

jpanikulam opened the pull request:
Buoy detection
### What this does
- Simple detection of buoys in 2d, Improvements to come @tessbianchi
- Adds a library of templated helper functions for PCL, and OpenCV/PCL interoperation
  - Adds a bunch of tools for projecting 2d segmentation into 3d
- Adds a tool for visualizing buoy detections in Rviz
- Adds a **very rough** first VisionRequest.srv, that is our first step towards the way missions will request vision information.
- Adds a library called backward.hpp that can be used to debug segfaults, pretty-printing stacktraces (Not worth reviewing, this is from [here.](https://github.com/bombela/backward-cpp))
### What this does not do
- All buoy colors, that's easy from here
- Add the actual buoy bumping mission, this is for another PR

Submitting to David for review.

jpanikulam committed:
TOOLS: Add library for stacktraces

- Add the famous Bombela backward.cpp/hpp
- This enables us to do stacktraces after segfaults, etc

jpanikulam committed:
PCL: Add sub8_pcl perception package

jpanikulam committed:
PCL: Add pcl 2d-3d reprojection utilities

jpanikulam committed:
PCL: Add simple buoy detector

- Still many false positives
TODO:
    - Proper kd-tree for color-based region-growing (HSV?)
    - std-dev/kurtosis estimation

jpanikulam committed:
PCL: Add simple normal-estimation

jpanikulam committed:
BUILD: Add PCL dependency

jpanikulam committed:
BUILD: Assume backward-cpp has bfd

jpanikulam committed:
PCL: Improve sub8_pcl library structure

jpanikulam committed:
PCL: Clean up CMakeLists

jpanikulam committed:
PCL: Add visualization of buoy detection

jpanikulam committed:
PCL: Add box segmentation

jpanikulam committed:
PCL: Double speed of ray projection (OMP)

- Running on two cores, speed is doubled
- With more cores, expect greater increase
- The bottleneck is reducing the list of distances to find the index
    - (OpenMP doesn't have a great way to natively do this)

jpanikulam committed:
PCL;CV: Add contour centroid utility function

jpanikulam committed:
PCL: Add pcl point -> Eigen-Vector conversion

Since this is templated, it should work for cv::Point3f as well.
TODO:
    - Can we template returns to make this usable for both Vector3f and
      Vector3d?

jpanikulam committed:
PCL: Add funcional box segmentation

jpanikulam committed:
PCL: Fix build

jpanikulam committed:
PCL: Add compile-time visualization control

jpanikulam committed:
PERCEPTION: Add VisionRequest service

jpanikulam committed:
PERCEPTION: Make buoy detector service-based

jpanikulam committed:
PCL: Move filters to pcl_tools

jpanikulam committed:
MERGE: Merge uf-mil/Sub8 into jpanikulam/buoys

Conflicts:
	utils/sub8_msgs/CMakeLists.txt

jpanikulam committed:
PCL: Add rviz visualization tool

jpanikulam committed:
PCL: Update readme

jpanikulam commented:
Note that 2,007 of the 2,948 lines in this PR are from backward.hpp, are external, and as such do not need to be reviewed.

DSsoto commented on gnc/perception/sub8_pcl/include/sub8_pcl/geometry.hpp:
```diff
@@ -0,0 +1,133 @@
+
+    Eigen::Vector3f cloud_pt(point.x, point.y, point.z);
+    Eigen::Vector3f difference = (line_pt - cloud_pt);
+    double distance = (difference - (difference.dot(direction) * direction)).norm();
```
Shouldnt this be `distance = direction.cross(difference).norm() / direction.norm()` ?
See [http://mathworld.wolfram.com/Point-LineDistance3-Dimensional.html](url), formula (9)
(might be equivalent)

jpanikulam replied:
Looks like they're equivalent: [implemented, on wikipedia](https://en.wikipedia.org/wiki/Distance_from_a_point_to_a_line#Vector_formulation), and [wolfram's formula](https://en.wikipedia.org/wiki/Distance_from_a_point_to_a_line#Another_vector_formulation), on wikipedia.

DSsoto commented on gnc/perception/sub8_pcl/nodes/buoy.cc:
```diff
@@ -0,0 +1,291 @@
+#define BACKWARD_HAS_BFD 1
+#include <sub8_build_tools/backward.hpp>
+
+#define VISUALIZE
```
Great idea! I'll steal this

DSsoto commented on gnc/perception/sub8_pcl/nodes/buoy.cc:
```diff
@@ -0,0 +1,291 @@
+  std::string tf_frame;
+  Eigen::Vector3f position;
+
+  if (!got_cloud) {
```
ROS_ERROR() ?

jpanikulam replied:
Added ROS_ERROR

DSsoto commented on gnc/perception/sub8_pcl/nodes/buoy.cc:
```diff
@@ -0,0 +1,291 @@
+    }
+    sub::PointXYZT centroid_projected =
+        sub::project_uv_to_cloud(*point_cloud_raw, contour_centroid, camera_model);
+    buoy_centers.push_back(sub::point_to_eigen(centroid_projected));
```
Name could be misleading since we expect these  3D points to be on the surface of the buoy even though they appear in the center in the image. I'm probably nitpicking here.

jpanikulam replied:
Good catch. I can see this causing problems, if we write missions with the assumption that the returned point is the center of the buoy. I've added a few lines to slide the point along the vector by one buoy radius, to get an approximate center

DSsoto commented on gnc/perception/sub8_pcl/nodes/buoy.cc:
```diff
@@ -0,0 +1,291 @@
+      closest_buoy_index = buoy_index;
+    }
+  }
+  center = buoy_centers[closest_buoy_index];
```
Logic: is the closest point necessarily the most likely to belong to the desired buoy? Maybe once you calculate a segmented point cloud you could use a heuristic taking into account the size of the cloud and the average distance and determine if it is indeed "buoy sized" and of the right color.

jpanikulam replied:
Good idea. Next version, we should do a combination of distance, deviation in size, and deviation in color.

We also ought to try point cloud geometry matching (RANSAC?) and try to track the thing for a few frames, instead of just the most recent, and shit the bed if we can't track

DSsoto commented on gnc/perception/sub8_pcl/nodes/buoy.cc:
```diff
@@ -0,0 +1,291 @@
+  center = buoy_centers[closest_buoy_index];
+}
+
+void Sub8BuoyDetector::image_callback(const sensor_msgs::ImageConstPtr &image_msg,
```
It seems like we are converting images constantly even in the absence of a VisionRequest service call. Maybe we should have a member variable indicating when there's an active request so we don't waste computations.

jpanikulam replied:
Fixed this - the image conversion is pretty negligible, but I benchmarked the cloud conversion, which clocked in at 160ms per callback. Moved that into the service call. 

It makes the visualization a little bit laggier, but since the tf cache will make it work in practice.

DSsoto commented on gnc/perception/sub8_pcl/nodes/buoy.cc:
```diff
@@ -0,0 +1,291 @@
+  image_time = image_msg->header.stamp;
+}
+
+void Sub8BuoyDetector::cloud_callback(const sensor_msgs::PointCloud2::ConstPtr &input_cloud) {
```
Again, it seems like we could skip point cloud processing in the absence of a VisionRequest service call. To deal with image and point cloud synchronization, we could wait until we get a pair with similar enough time-stamps after a request is active and then try to locate the buoy.

DSsoto commented:
Everything looks pretty good, impressive work! I would have tried to run it but I don't have any buoy bags. 

Since most of our perception will be done using PCL and OpenCV, I suggest maybe having sub8_pcl become sub8_perception or something of the sort. That way we could include the torpedo vision and other vision routines as services or actions along with your buoy stuff and share the infrastructure (especially the pcl code you wrote which would be nice to have available from my end as well). Otherwise the repo will get bloated with pretty much identical copies of a lot of the files in this PR for other perception packages.

Lastly, I was wondering why you didn't make buoy detection an action instead of a service so we could get feedback and stop it prematurely if necessary.

jpanikulam committed:
PCL: Address PR #63 comments

jpanikulam commented:
I agree with your suggestion to rename the package, just did that in the final commit (Might have broken the build, bear with me).

I didn't make it an action because there's not much intermediate feedback, really we either detect the buoy and get a centroid, or we get no results at all. I suspect most vision tasks would involve repeated calls to a service instead of an action. Maybe actions will be more appealing after we do testing.

Stand-by for commits + merging master in

jpanikulam committed:
PCL: Rename sub8_pcl to sub8_perception

jpanikulam committed:
MERGE: uf-mil/master into jpanikulam/buoys

    - Bringing our branch up to date

jpanikulam commented:
Addressed comments in [this commit](https://github.com/jpanikulam/Sub8/commit/c22bbc4ba3e39c7cbbac473db968c0a5c82c6b79), and also renamed the package. Will merge pending final review.

Great comments, btw
